packages/ai-jsx/src/lib/asr/asr.tsx

```diff
@@ -0,0 +1,428 @@
+    const view = new DataView(pcmBuffer);
+    let index = 0;
+    let energy = 0.0;
+    buffers.forEach((buffer) => {
+      buffer.forEach((sample) => {
+        energy += sample * sample;
+        const i16 = Math.max(
+          -32768,
+          Math.min(32767, Math.floor(sample * 32768)),
+        );
+        view.setInt16(index, i16, true);
+        index += 2;
+      });
+      this.updateVad(energy / (index / 2));
+    });
+    return pcmBuffer;
+  }
+  private updateVad(energy: number) {
+    const dbfs = 10 * Math.log10(energy);
+    if (dbfs < -50) {
+      this.numSilentFrames++;
+      if (this.numSilentFrames == 100) {
+        this.dispatchEvent(new CustomEvent('vad', { detail: false }));
+      }
+    } else {
+      if (this.numSilentFrames >= 100) {
+        this.dispatchEvent(new CustomEvent('vad', { detail: true }));
+      }
+      this.numSilentFrames = 0;
+    }
+  }
+}
+
+export class SpeechRecognitionBase extends EventTargetImpl {
+  private lastUtteranceEndTime: number;
+  protected manager: MicManager;
+  protected socket?: WebSocket;
+  protected name: string;
+  protected language?: string;
+  private outBuf: ArrayBuffer[];
+  constructor(manager: MicManager, name: string, language?: string) {
+    super();
+    this.manager = manager;
+    this.name = name;
+    this.language = language;
+    this.lastUtteranceEndTime = 0;
+    this.outBuf = [];
+  }
+  async start() {}
+  close() {
+    if (this.socket) {
+      this.sendClose();
+      this.socket.close();
+    }
+  }
+  protected async fetchToken() {
+    const response = await fetch('/asr/api', {
+      method: 'POST',
+      body: JSON.stringify({ provider: this.name }),
+    });
+    const json = await response.json();
+    return json.token;
+  }
+  protected startInternal(url: string, protocols?: string[]) {
+    const startTime = performance.now();
+    console.log(`${this.name} socket connecting...`);
+    this.outBuf = [];
+    this.socket = new WebSocket(url, protocols);
+    this.socket.binaryType = 'arraybuffer';
+    this.socket.onopen = (event) => {
+      const elapsed = performance.now() - startTime;
+      console.log(`${this.name} socket opened, et=${elapsed.toFixed(0)}`);
+      this.handleOpen();
+      this.flush();
+    };
+    this.socket.onmessage = (event) => {
+      const result = JSON.parse(event.data);
+      this.handleMessage(result);
+    };
+    this.socket.onerror = (event) => {
+      console.log(`${this.name} socket error`);
+    };
+    this.socket.onclose = (event) => {
+      console.log(
+        `${this.name} socket closed, code=${event.code} reason=${event.reason}`,
+      );
+    };
+    this.manager.addEventListener('chunk', (event: CustomEvent) => {
+      if (this.socket!.readyState == 1) {
+        this.sendChunk(event.detail);
+      } else if (this.socket!.readyState == 0) {
+        this.outBuf.push(event.detail);
+      } else {
+        console.warn(`${this.name} socket closed`);
```

Comment:
Maybe this should be an error instead of a warning?

Reply:
I can get behind that. Probably should auto-reconnect, although it will need a new token, so I need to think about that, but... next PR